File: go/normalize.go
package wh40kdc

import (
	"regexp"
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

// Name normalization for diacritic- and punctuation-insensitive lookup.
//
// Go mirror of the TypeScript normalizeName (tools/src/data/normalize.ts),
// the Rust normalize_name, and the Python normalize_name; pinned together by
// the shared conformance/normalize.json corpus.

// quotesRe matches the apostrophe and quote variants stripped during
// normalization: ' U+0027, ’ U+2019, ‘ U+2018, ` U+0060, " U+0022, “ U+201C,
// ” U+201D. (Deliberately the same seven the other ports strip — e.g. ‛ U+201B
// is NOT included, so it passes through.)
var quotesRe = regexp.MustCompile("['’‘`\"“”]")

// spaceHyphenRe collapses any run of whitespace or ASCII hyphen to a single
// space. \s + \p{Z} + vertical tab covers the Unicode whitespace the corpus
// and fuzz pool exercise (incl. NBSP U+00A0 and ideographic space U+3000);
// only the ASCII hyphen-minus is treated as a hyphen (other hyphen variants
// pass through, matching the reference impls).
var spaceHyphenRe = regexp.MustCompile(`[\s\p{Z}\x{000b}-]+`)

// leadingTheRe matches a leading "The " (case-insensitive, after trim) and
// captures the remainder. Mirror of the TS `stripLeadingThe` regex
// `/^the\s+(.+)$/i`.
var leadingTheRe = regexp.MustCompile(`(?i)^the\s+(.+)$`)

// NormalizeName reduces a display name to a canonical lookup key. The
// transform, in order: NFD-decompose → strip combining marks → lowercase →
// remove quote variants → collapse whitespace/hyphen runs to a single space →
// trim. The result is for comparison only, not display.
func NormalizeName(input string) string {
	decomposed := norm.NFD.String(input)
	// Strip marks *before* lowercasing — load-bearing for İ (U+0130), whose
	// lowercase form introduces a combining dot that must survive.
	var b strings.Builder
	b.Grow(len(decomposed))
	for _, r := range decomposed {
		if !unicode.Is(unicode.M, r) {
			b.WriteRune(r)
		}
	}
	lowered := strings.ToLower(b.String())
	noQuotes := quotesRe.ReplaceAllString(lowered, "")
	collapsed := spaceHyphenRe.ReplaceAllString(noQuotes, " ")
	return strings.TrimSpace(collapsed)
}

// StripLeadingThe strips a leading "The " (case-insensitive, after trimming)
// from a display name, returning the remainder and true, or ("", false) when
// there is no leading "The " to strip.
//
// Used only by roster import to bridge the leading-article mismatch between
// data names and roster exports in BOTH directions ("The Bloody Twins" ↔
// "Bloody Twins", "Fire Axe" ↔ "The Fire Axe"). Deliberately NOT folded into
// NormalizeName, whose key is shared by unit, faction, and ability lookup,
// where dropping a leading "The" would collide distinct entities (e.g. "The
// Emperor's Champion"). Mirror of the TS stripLeadingThe.
func StripLeadingThe(input string) (string, bool) {
	m := leadingTheRe.FindStringSubmatch(strings.TrimSpace(input))
	if m == nil {
		return "", false
	}
	return strings.TrimSpace(m[1]), true
}

File: crates/wh40kdc/src/data/normalize.rs
//! Name normalization for diacritic- and punctuation-insensitive lookup.
//!
//! Warhammer 40,000 is played globally and many entity names carry diacritics or
//! punctuation — "Khârn the Betrayer", "T'au", "Be'lakor". A user typing the
//! plain-ASCII form of a name must still find the entity. Every name comparison
//! in this crate routes through [`normalize_name`] so the matching rule is
//! defined in exactly one place; it is exported so consumers can reproduce the
//! same behaviour in their own search UIs.
//!
//! This is the Rust mirror of the TypeScript `normalizeName`
//! (`tools/src/data/normalize.ts`); the two implementations are pinned together
//! by the shared `conformance/normalize.json` corpus.

use unicode_normalization::{char::is_combining_mark, UnicodeNormalization};

/// Reduce a display name to a canonical lookup key.
///
/// The transform, in order:
/// 1. Unicode NFD-decompose, then strip combining marks — `Khârn` → `Kharn`.
/// 2. Casefold to lower case.
/// 3. Remove apostrophe and quote variants (`'`, `’`, `‘`, `` ` ``, `"`, `“`,
///    `”`) — `T'au` → `Tau`.
/// 4. Collapse any run of whitespace or hyphens to a single space, then trim —
///    `Be'lakor` → `belakor`, `the   betrayer` → `the betrayer`.
///
/// The result is intended only for comparison; it is not a display value.
///
/// # Examples
///
/// ```
/// use wh40kdc::normalize_name;
///
/// assert_eq!(normalize_name("Khârn the Betrayer"), "kharn the betrayer");
/// assert_eq!(normalize_name("T'au"), "tau");
/// assert_eq!(normalize_name("Be'lakor"), "belakor");
/// assert_eq!(normalize_name("  Adeptus   Astartes  "), "adeptus astartes");
/// ```
pub fn normalize_name(input: &str) -> String {
    // NFD-decompose, drop combining marks, then lowercase. Ordering matches the
    // TS `.normalize("NFD").replace(/\p{Diacritic}/gu, "").toLowerCase()`.
    let decomposed: String = input.nfd().filter(|c| !is_combining_mark(*c)).collect();

    let mut out = String::with_capacity(decomposed.len());
    let mut pending_space = false;
    for c in decomposed.chars().flat_map(char::to_lowercase) {
        match c {
            // Apostrophe / quote variants are removed outright (not spaced),
            // so `T'au` → `tau` rather than `t au`.
            '\'' | '\u{2019}' | '\u{2018}' | '`' | '"' | '\u{201C}' | '\u{201D}' => {}
            // Whitespace and hyphens collapse to a single separating space; a
            // leading run is suppressed (out is still empty) and a trailing run
            // is dropped because the space is only flushed before the next char.
            _ if c.is_whitespace() || c == '-' => {
                if !out.is_empty() {
                    pending_space = true;
                }
            }
            _ => {
                if pending_space {
                    out.push(' ');
                    pending_space = false;
                }
                out.push(c);
            }
        }
    }
    out
}

/// Strip a leading "The " (case-insensitive, after trimming) from a display
/// name, returning the remainder. Returns `None` when there is no leading
/// "The " to strip.
///
/// Used only by roster import to bridge the leading-article mismatch between
/// data names and roster exports in BOTH directions ("The Bloody Twins" ↔
/// "Bloody Twins", "Fire Axe" ↔ "The Fire Axe"). Deliberately NOT folded into
/// [`normalize_name`]: that key is shared by unit, faction, and ability lookup,
/// where dropping a leading "The" would collide distinct entities (e.g. "The
/// Emperor's Champion"). Mirror of the TS `stripLeadingThe`.
pub fn strip_leading_the(input: &str) -> Option<String> {
    let trimmed = input.trim();
    // Byte length of the first three characters, so the prefix slice is valid
    // even if they are multi-byte (no ASCII assumption).
    let mut prefix_end = 0;
    let mut chars = trimmed.chars();
    for _ in 0..3 {
        match chars.next() {
            Some(c) => prefix_end += c.len_utf8(),
            None => return None,
        }
    }
    if !trimmed[..prefix_end].eq_ignore_ascii_case("the") {
        return None;
    }
    let after = &trimmed[prefix_end..];
    let rest = after.trim_start();
    // Require at least one stripped whitespace char (so "theatre" is rejected)
    // and a non-empty remainder.
    if rest.len() == after.len() || rest.is_empty() {
        return None;
    }
    Some(rest.to_string())
}

#[cfg(test)]
mod tests {
    use super::normalize_name;

    #[test]
    fn strips_diacritics() {
        assert_eq!(normalize_name("Khârn the Betrayer"), "kharn the betrayer");
        assert_eq!(normalize_name("Magnús"), "magnus");
    }

    #[test]
    fn removes_quote_variants() {
        assert_eq!(normalize_name("T'au"), "tau");
        assert_eq!(normalize_name("Be'lakor"), "belakor");
        assert_eq!(normalize_name("\u{2018}quoted\u{2019}"), "quoted");
    }

    #[test]
    fn collapses_whitespace_and_hyphens() {
        assert_eq!(normalize_name("the   betrayer"), "the betrayer");
        assert_eq!(normalize_name("space--marines"), "space marines");
        assert_eq!(
            normalize_name("  leading and trailing  "),
            "leading and trailing"
        );
    }

    #[test]
    fn distinct_names_stay_distinct() {
        assert_ne!(normalize_name("Khorne"), normalize_name("Khârn"));
    }
}
